llama_index/llms/huggingface.py

```diff
@@ -294,3 +311,243 @@ def stream_chat(
+            kwargs: See the class-level Fields.
+        """
+        super().__init__(**kwargs)  # Populate pydantic Fields
+        try:
+            from huggingface_hub import AsyncInferenceClient, InferenceClient
+        except ModuleNotFoundError as exc:
+            raise ImportError(
+                f"{type(self).__name__} requires huggingface_hub with its inference"
+                " extras, please run `pip install huggingface_hub[inference]`."
+            ) from exc
+        self._sync_client = InferenceClient(**self._get_inference_client_kwargs())
+        self._async_client = AsyncInferenceClient(**self._get_inference_client_kwargs())
+
+    def validate_supported(self, task: str) -> None:
+        """
+        Confirm the contained model_name is deployed on the Inference API service.
+
+        Args:
+            task: Hugging Face task to check within. A list of all tasks can be
+                found here: https://huggingface.co/tasks
+        """
+        all_models = self._sync_client.list_deployed_models(frameworks="all")
+        try:
+            if self.model_name not in all_models[task]:
+                raise ValueError(
+                    f"The Inference API service doesn't have the model"
+                    f" {self.model_name!r} deployed."
+                )
+        except KeyError as exc:
+            raise KeyError(
+                f"Input task {task!r} not in possible tasks {list(all_models.keys())}."
+            ) from exc
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        raise NotImplementedError
+
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
+        return conversational_output_to_chat_response(
+            output=self._sync_client.conversational(
+                **{**chat_messages_to_conversational_kwargs(messages), **kwargs}
+            )
+        )
+
+    def complete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
+        return CompletionResponse(
+            text=self._sync_client.text_generation(prompt, **kwargs)
+        )
+
+    def stream_chat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponseGen:
+        raise NotImplementedError
+
+    def stream_complete(self, prompt: str, **kwargs: Any) -> CompletionResponseGen:
+        raise NotImplementedError
+
+    async def achat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponse:
+        raise NotImplementedError
+
+    async def acomplete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
+        raise NotImplementedError
+
+    async def astream_chat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponseAsyncGen:
+        raise NotImplementedError
+
+    async def astream_complete(
+        self, prompt: str, **kwargs: Any
+    ) -> CompletionResponseAsyncGen:
+        raise NotImplementedError
+
+    def embed_documents(self, texts: List[str]) -> List[Embedding]:
```

Comment:
I think there are several methods on the embeddings base class not implemented here -- this will likely cause errors?

Tbh I'd prefer to have a separate embeddings implementation. The only duplicate code is the client setup, which could be a util file somewhere?

Reply:
I see that as leading to unnecessary proliferation of types and extra imports, it also spreads things out which makes the code more arduous to comprehend. I think the current solution I have, a name alias in `embeddings/hugging_face.py` is simplest and leads to equally readable client code.

To share, promptfoo tackles this with embeddings done in a subclass: https://github.com/promptfoo/promptfoo/blob/0.25.0/src/providers/ollama.ts

I can just subclass and do the mixing in of `Embeddings` within `embeddings/`, since separation of types/behaviors is the preference here

Reply:
Okay, split them up, and I also correspondingly split up the test cases